### PR TITLE
feat(api): incorporate risk-adjusted metrics as primary perps ranking mechanism (calmar ratio)

### PR DESCRIPTION
**Pull request by mzkrasner**
# Perps Competition Risk-Adjusted Ranking with Calmar Ratio

## Overview

This PR introduces risk-adjusted performance rankings for perpetual futures competitions using the **Calmar Ratio** as the primary ranking metric. The implementation provides fair competition rankings by evaluating both returns and risk, with special handling to prevent gaming through mid-competition transfers.

### Calmar Ratio Implementation
- **Primary ranking metric** for perps competitions
- Formula: `Annualized Return / |Max Drawdown|`
- Rewards consistent performance with controlled risk
- Penalizes high-volatility strategies appropriately

### Transfer Prohibition & Violation Detection
- **Mid-competition transfers are strictly prohibited**
- All deposits/withdrawals during competition are flagged as violations
- Transfer history is stored for audit and admin review
- New admin endpoint to query transfer violations by competition

### Simple Return Calculation
- Uses portfolio snapshots to calculate `(endValue/startValue) - 1`
- No complex time-weighted calculations needed
- Clean and transparent performance measurement

#### New Tables
- `perps_risk_metrics`: Stores Calmar ratio, simple returns, and max drawdown
- `perps_transfer_history`: Tracks all transfers for violation detection

#### Removed Tables
- ~~`perps_twr_periods`~~ (No longer needed without TWR)

#### New Services
- **CalmarRatioService**: Calculates risk metrics using portfolio snapshots
  - Simple return calculation from first/last snapshots
  - Max drawdown via efficient SQL queries
  - Annualized return computation

#### Updated Services
- **PerpsMonitoringService**: 
  - Flags ALL mid-competition transfers as violations
  - Stores transfer history for audit purposes
- **PerpsDataProcessor**:
  - Integrated Calmar ratio calculations
  - Circuit breaker for error handling
  - Batch processing optimizations

#### New Admin Endpoint
```
GET /api/admin/competition/{competitionId}/transfer-violations
```
Returns agents with transfer violations:
```json
{
  "success": true,
  "violations": [
    {
      "agentId": "...",
      "agentName": "...",
      "transferCount": 2
    }
  ]
}
```

#### Updated Endpoints
All competition and leaderboard endpoints now include risk metrics:
- `calmarRatio`: Risk-adjusted performance score
- `simpleReturn`: Basic return percentage
- `maxDrawdown`: Maximum portfolio decline
- `hasRiskMetrics`: Indicates if metrics are calculated

### SQL-First Approach
- All aggregations pushed to database level
- Efficient `GROUP BY` and `COUNT` for transfer violations
- Lateral joins for leaderboard queries
- Compound indexes on frequently queried columns

### Batch Processing
- Risk metrics calculated in parallel batches
- Circuit breaker prevents cascading failures
- Error capping limits memory usage

### Comprehensive Test Coverage
- Unit tests for CalmarRatioService
- E2E tests for transfer violation detection
- Mock Symphony server configured with test transfers
- Edge cases: zero drawdown, negative returns, short competitions

### Test Scenarios
✅ Calmar ratio calculation with various market conditions
✅ Transfer violation detection and reporting
✅ Leaderboard ranking with risk metrics
✅ Admin endpoint authorization and error handling

### For Running Competitions
1. Risk metrics will be calculated on next cron run
2. Rankings will automatically switch to Calmar-based
3. Historical data preserved but not recalculated

### For New Competitions
1. Participants must be informed: **No transfers allowed during competition**
2. Initial capital is locked at competition start
3. Performance measured purely on trading skill

### Transfer Violation Detection
- Admin dashboard can query violations per competition
- Automated alerts for detected violations
- Full audit trail with transfer details

### Risk Metrics Dashboard
- Real-time Calmar ratio updates
- Max drawdown tracking
- Simple return calculations
- Portfolio volatility analysis

## Benefits

1. **Fair Competition**: Eliminates advantage from capital injections
2. **Risk-Aware Ranking**: Rewards sustainable strategies over gambling
3. **Transparent Metrics**: Simple returns are easy to understand
4. **Scalable Architecture**: SQL aggregations handle large competitions
5. **Admin Control**: Full visibility into transfer violations

## Breaking Changes

- ⚠️ **Mid-competition transfers now prohibited** - participants must be notified
- API responses include new risk metric fields (backwards compatible)
- Leaderboard ranking algorithm changed to Calmar-based

## Performance Impact

- Minimal: Risk calculations run during existing cron jobs
- SQL-optimized queries with proper indexes
- Batch processing prevents memory issues
- Circuit breaker prevents runaway calculations

## Security Considerations

- Transfer detection cannot be gamed (all transfers flagged)
- Admin-only access to violation data
- Encrypted storage of sensitive metrics
- Rate limiting on all endpoints

## Future Enhancements

- [ ] Sharpe Ratio as additional metric
- [ ] Sortino Ratio for downside risk
- [ ] Custom risk metrics per competition
- [ ] Automated violation penalties
- [ ] Risk metric history graphs

## Documentation

- Updated OpenAPI specifications
- Admin guide for violation handling
- Participant rules clearly state transfer prohibition
- Technical docs for risk metric calculations

---

## Checklist

- [x] Database migrations created and tested
- [x] Unit tests passing (100% coverage on new code)
- [x] E2E tests passing
- [x] Linting and type checking clean
- [x] Performance testing completed
- [x] Admin endpoint documented
- [x] Breaking changes communicated
- [x] Security review completed

### CODE REVIEW

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts [hunk 1]
```diff
@@ -130,7 +139,7 @@ export class SymphonyPerpsProvider implements IPerpsDataProvider {
 
   constructor(apiUrl?: string) {
     // Use provided URL or fall back to config/environment
-    this.baseUrl = apiUrl || "https://api.symphony.finance";
+    this.baseUrl = apiUrl || "https://api.symphony.io";
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/perps/symphony-perps.provider.ts:135]: Ignore.

File: apps/api/src/services/calmar-ratio.service.ts [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    if (daysInPeriod < 1) {
+      return periodReturn;
+    }
+
+    const yearsInPeriod = daysInPeriod / this.DAYS_PER_YEAR;
+
+    // Use Decimal for precise calculation
+    const onePlusReturn = new Decimal(1).plus(periodReturn);
+    const exponent = new Decimal(1).div(yearsInPeriod);
+    const annualized = onePlusReturn.pow(exponent).minus(1);
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/calmar-ratio.service.ts:194]: Addressed in [f3263ec](https://github.com/recallnet/js-recall/pull/1273/commits/f3263ecae10201f6b9179c675fbaad49796842b0)

File: apps/api/src/services/perps-data-processor.service.ts [hunk 3]
```diff
@@ -749,6 +767,144 @@ export class PerpsDataProcessor {
...
+
+      batchResults.forEach((result, idx) => {
+        if (result.status === "fulfilled") {
+          results.successful++;
+        } else {
+          results.failed++;
+          batchFailures++;
+
+          const agent = batch[idx];
+          const errorMsg = `Agent ${agent?.agentId}: ${result.reason?.message || "Unknown error"}`;
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/perps-data-processor.service.ts]: Addressed in [f3263ec](https://github.com/recallnet/js-recall/pull/1273/commits/f3263ecae10201f6b9179c675fbaad49796842b0)

File: packages/db/src/repositories/perps.ts [hunk 4]
```diff
@@ -734,6 +744,117 @@ export class PerpsRepository {
...
+          agentId: competitionAgents.agentId,
+        })
+        .from(competitionAgents)
+        .where(
+          and(
+            eq(competitionAgents.competitionId, competitionId),
+            eq(competitionAgents.status, "active"),
+          ),
+        )
+        .as("active_agents");
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply packages/db/src/repositories/perps.ts]: Updated in [3b4d032](https://github.com/recallnet/js-recall/pull/1273/commits/3b4d0320b333a14aaf85a1e0ab196d2b9199ae13)

File: apps/api/src/services/perps-monitoring.service.ts [hunk 5]
```diff
@@ -532,4 +543,49 @@ export class PerpsMonitoringService {
...
+        (t) => ({
+          agentId,
+          competitionId,
+          type: t.type,
+          amount: t.amount.toString(),
+          asset: t.asset,
+          fromAddress: t.from,
+          toAddress: t.to,
+          txHash:
+            t.txHash || `${t.timestamp.toISOString()}-${t.type}-${t.amount}`, // Fallback for missing txHash
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/perps-monitoring.service.ts]: Fixed in [a3cff1](https://github.com/recallnet/js-recall/pull/1273/commits/1a3cff189f5bbf3353295ba2dc90752bac961087)

File: apps/api/src/services/calmar-ratio.service.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    if (Math.abs(yearsInPeriod - 1) < 0.0001) {
+      return periodReturn;
+    }
+
+    // Use Decimal for precise calculation
+    const onePlusReturn = new Decimal(1).plus(periodReturn);
+    const exponent = new Decimal(1).div(yearsInPeriod);
+    const annualized = onePlusReturn.pow(exponent).minus(1);
+
+    return annualized.toNumber();
```
Comment (smrz2001) [inline apps/api/src/services/calmar-ratio.service.ts]: The `.toNumber()` could lose the precision that we get from `Decimal`.

Can just return the `Decimal` directly? Happens in a few places, can update them all.

File: apps/api/src/services/competition.service.ts [hunk 7]
```diff
@@ -2882,4 +2998,74 @@ export class CompetitionService {
...
+
+      // Create a map for quick lookup
+      const agentMap = new Map(agents.map((agent) => [agent.id, agent.name]));
+
+      // 4. Combine data and return
+      const results = violationCounts.map((violation) => ({
+        agentId: violation.agentId,
+        agentName: agentMap.get(violation.agentId) ?? "Unknown Agent",
+        transferCount: violation.transferCount,
+      }));
```
Comment (smrz2001) [inline apps/api/src/services/competition.service.ts]: If you join with the agents table and include the agent name in the select in `getCompetitionTransferViolationCounts()`, this code can be eliminated.

File: apps/api/src/services/perps-data-processor.service.ts [hunk 8]
```diff
@@ -749,6 +767,150 @@ export class PerpsDataProcessor {
...
+              `Circuit breaker triggered after ${consecutiveFailures} consecutive batch failures. ` +
+                `${remainingAgents} agents skipped.`,
+            );
+          }
+
+          break; // Exit the loop
+        }
+      } else {
+        // Reset counter if batch had any successes
+        consecutiveFailures = 0;
```
Comment (smrz2001) [inline apps/api/src/services/perps-data-processor.service.ts]: Processing will continue even through back-to-back batches with a 90% failure rate. Do we want the circuit break to work this way?

File: apps/api/src/services/calmar-ratio.service.ts (hunk 6 above)
Comment (mzkrasner) [reply apps/api/src/services/calmar-ratio.service.ts]: Good call - updated in [38a0857](https://github.com/recallnet/js-recall/pull/1273/commits/38a085733bcb836b0008cf2e02e9572b0e78fe76) to avoid the potential precision loss between internal calculations

File: apps/api/src/services/competition.service.ts (hunk 7 above)
Comment (mzkrasner) [reply apps/api/src/services/competition.service.ts]: Nice - great optimization - added in [c55faf2](https://github.com/recallnet/js-recall/pull/1273/commits/c55faf282f7863e6dc95494a7fd2f16712f47c76)

File: apps/api/src/services/perps-data-processor.service.ts (hunk 8 above)
Comment (mzkrasner) [reply apps/api/src/services/perps-data-processor.service.ts]: I ended up altering this in [82ed3c4](https://github.com/recallnet/js-recall/pull/1273/commits/82ed3c4b022a3c57453c67a2c8cb45d4e612a16d) to conform to a hybrid approach:

1. **Added retry logic** (3 retries with exponential backoff: 1s, 2s, 4s) - same pattern as portfolio snapshotter
2. **Circuit breaker now monitors but doesn't stop** - triggers alerts at 80% failure rate for visibility
3. **All agents still get processed** - ensures rankings remain as complete as possible

This alters the circuit breaker to now serve as an alerting mechanism rather than a stopping mechanism. LMK what you think!

File: apps/api/src/services/perps-data-processor.service.ts [hunk 9]
```diff
@@ -749,6 +767,150 @@ export class PerpsDataProcessor {
...
+      });
+
+      // Circuit breaker: Check if entire batch failed
+      const batchFailureRate = batchFailures / batch.length;
+      if (batchFailureRate === 1) {
+        consecutiveFailures++;
+
+        if (consecutiveFailures >= MAX_CONSECUTIVE_FAILURES) {
+          systemicFailureDetected = true;
+          const remainingAgents = successfulAgents.length - (i + BATCH_SIZE);
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/perps-data-processor.service.ts]: Also addressed in [82ed3c4](https://github.com/recallnet/js-recall/pull/1273/commits/82ed3c4b022a3c57453c67a2c8cb45d4e612a16d)

File: apps/api/src/services/calmar-ratio.service.ts [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        `[CalmarRatio] Simple return calculated: ${(simpleReturn.toNumber() * 100).toFixed(4)}% (${startValue.toFixed(2)} → ${endValue.toFixed(2)})`,
+      );
+
+      // 3. Calculate Max Drawdown using SQL
+      const maxDrawdown = await calculateMaxDrawdownSQL(
+        agentId,
+        competitionId,
+        startDate,
+        endDate,
+      );
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/calmar-ratio.service.ts:105]: This is a substantial finding - amended in [41a955a](https://github.com/recallnet/js-recall/pull/1273/commits/41a955a5ab3475d4ee244cefa5b105edeea35670)

File: apps/api/src/services/perps-data-processor.service.ts [hunk 11]
```diff
@@ -749,6 +767,209 @@ export class PerpsDataProcessor {
...
+      // Circuit breaker monitoring (but don't stop processing)
+      const batchFailureRate = batchFailures / batch.length;
+      if (batchFailureRate >= CIRCUIT_BREAKER_THRESHOLD) {
+        failedBatches++;
+
+        // Calculate overall failure rate
+        const overallFailureRate = failedBatches / totalBatches;
+
+        // Log warning if we're seeing systemic issues
+        if (overallFailureRate >= CIRCUIT_BREAKER_THRESHOLD) {
```
Comment (smrz2001) [inline apps/api/src/services/perps-data-processor.service.ts]: Rewritten logic is great!

Just one minor comment -  batch-level failures and systemic failures are distinct failure patterns that should have distinct thresholds. Using the same 0.8 threshold for both might be problematic.

I don't have a strong opinion on the systemic failure threshold. Maybe 50%? Or 25%? :shrug:

Comment (mzkrasner) [reply apps/api/src/services/perps-data-processor.service.ts]: Ah good call - set this to 50% in [d0a6b07](https://github.com/recallnet/js-recall/pull/1273/commits/d0a6b078212048120c2b4c7424680663d9c46eaa) (referenced as default in Netflix Hystrix, AWS best practices, and Azure)
